src/poprox_recommender/components/diversifiers/locality_calibration.py

```diff
@@ -25,13 +47,234 @@ def __call__(self, candidate_articles: ArticleSet, interest_profile: InterestPro
+        frequency_penalty=frequency_penalty,
+        model="gpt-4o-mini",
+    )
+    return chat_completion.choices[0].message.content
+
+
+def news_factor(headline, slug):
+    news = f"headline: {headline}\nslug: {slug}"
+    system_prompt = (
+        "Based on the given news information, summarize what topics that the news is related to."
+        "Each news article is related to 1-3 topics and each topic shuold not exceed five words."
+    )
+    return gpt_generate(system_prompt, news)
+
+
+# TODO: check backward or forward for past k articles
+def user_interest_generate(past_articles: Article, past_k: int):
+    past_article_infor = []
+    past_count = 0
+    for article in past_articles.articles:
+        past_count += 1
+        if past_count == past_k:
+            break
+        temp_article = {}
+        temp_article["headline"] = article.headline
+        temp_article["subhead"] = article.subhead
+        temp_article["categories"] = set([mention.entity.name for mention in article.mentions])
+        temp_article["topics"] = news_factor(article.headline, article.subhead)
+        past_article_infor.append(temp_article)
+
+    system_prompt = (
+        "You are asked to describe user interest based on his/her browsed news list."
+        " User interest includes the news [categories] and news [topics]"
+        " (under each [category] that users are interested in."
+    )
+
+    return gpt_generate(system_prompt, f"{past_article_infor}")
+
+
+def generate_narrative(news_list, user_interests):
+    system_prompt = (
+        "You are a personalized text generator."
+        " First, i will provide you with a news list that"
+        " includes both the [main news] and [related news]."
+        " Second, I will provide you with user interests,"
+        " including the [categories] of news that the user is interested in."
+        " Based on the input news list and user interests,"
+        " you are required to generate a personalized news summary centered around the [main news]."
+    )
+
+    input_prompt = "News List: \n" + f"{news_list}" + "User Interest: \n" + user_interests
+    return gpt_generate(system_prompt, input_prompt)
+
+
+model = SentenceTransformer("all-MiniLM-L6-v2")
+
+
+def related_article(current_news_slug, current_topics, clicked_articles, k):
+    # TODO: do we need to consider the cold start?
+    target_embedding = model.encode([current_news_slug])[0]
+
+    subject_candidate_headlines = []
+    for idx in range(len(clicked_articles)):
+        slug = clicked_articles[int(idx)].subhead
+        # category = set([mention.entity.name for mention in clicked_articles[int(idx)].mentions])
+
+        # if any(item in current_topics for item in category):
+        subject_candidate_headlines.append(slug)
+
+    candidate_embeddings = model.encode(subject_candidate_headlines)
+    similarities = cosine_similarity([target_embedding], candidate_embeddings)[0]
+    top_k_indices = np.argsort(similarities)[-k:][::-1]
```

Comment:
Here's what I'm thinking if we can fuse chronological order into candidate selection